charms/istio-pilot/src/charm.py

```diff
@@ -878,47 +852,63 @@ def _use_https(self) -> bool:
         return self._use_https_with_tls_provider() or self._use_https_with_tls_secret()
 
     def _use_https_with_tls_secret(self) -> bool:
-        """Return True if both SSL key and crt are set with save-tls-secret, False otherwise.
+        """Return True if tls-secret-id has been configured, False otherwise.
 
         Raises:
-            ErrorWithStatus: if one of the values is missing.
+            ErrorWithStatus("tls-secret-id was provided, but the secret could not be found - ...):
+                if a secret ID is passed through the tls-secret-id configuration option, but the
+                secret cannot be found.
+            ErrorWithStatus("Access to the istio-tls-secret must be granted."):
+                if the secret ID is passed, but the access to the secret is not granted.
+            ErrorWithStatus(Missing TLS value(s), please add them to the secret")
+                if any of the TLS values are missing from the secret.
         """
 
-        # Ensure the secret that holds the values exist otherwise fail as this
-        # is an error in our code
+        # Check if the tls-secret-id holds any value and get the secret
+        if not self._tls_secret_id:
+            return False
+
         try:
-            secret = self.model.get_secret(label=TLS_SECRET_LABEL)
+            secret = self.model.get_secret(id=self._tls_secret_id, label=TLS_SECRET_LABEL)
         except SecretNotFoundError:
-            return False
+            raise ErrorWithStatus(
+                "tls-secret-id was provided, but the secret could not be found - "
+                "please provide a secret ID of a secret that exists.\n"
+                f"Also make sure the name of the secret is {TLS_SECRET_LABEL}",
+                BlockedStatus,
+            )
+        except ModelError as e:
+            if "ERROR permission denied" in e.args:
+                # Block the unit when there is an ERROR permission denied
+                raise ErrorWithStatus(
+                    "Access to the istio-tls-secret must be granted.",
+                    BlockedStatus,
+                )
```

Comment:
Can we make this more actionable?  Maybe suggest to see `juju grant-secret --help` for more details?  I'm thinking of helping a user that doesn't really know about how secrets should work.  

"Permission denied trying to access `istio-tls-secret` (id: <secret-id>).  See `juju grant-secret --help` for details on granting permission"?  Its long though.  Could also write something short here telling them to see the logs for more instruction

Reply:
I can expand the message (: good suggestion

Reply:
Updated, please let me know what you think.

Reply:
Resolving as I have accepted the suggestion